WireAPI/Sources/WireAPI/Models/Backend/ServerTrustValidator.swift

```diff
@@ -0,0 +1,132 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+@preconcurrency import Security
+
+struct ServerTrustValidator: Sendable {
+
+    enum Error: Swift.Error {
+        case cannotLoadCertificate
+        case cannotRetrievePublicKey
+        case evaluatingServerTrustFailed((any Swift.Error)?)
+        case noPublicKeyOnServerTrust
+        case noMatchingPublicKey
+    }
+
+    private let pinnedKeys: [PinnedKey]
+
+    init(pinnedKeys: [PinnedKey]) {
+        self.pinnedKeys = pinnedKeys
+    }
+
+    /// Verifies the server `trust` for the given `host`.
+    ///
+    /// - Parameter trust: The `SecTrust` of the server.
+    /// - Parameter host: The host of the server.
+    /// - Throws: An error if server certificate should not be trusted.
+    /// - Note: If no pinned keys are found for the `host`, the server certificate is trusted.
+
+    func validate(trust: SecTrust, host: String) async throws {
+        let matches = pinnedKeys.filter { $0.matches(host: host) }
+
+        // If no keys are pinned for `host`, we trust the server certificate
+        guard !matches.isEmpty else { return }
```

Comment:
This is a change of behavior from  the `ServerCertificateTrust`. `ServerCertificateTrust` first verifies that a server certificate is valid before performing this check. However as we are only supposed to be validating certs from servers we control, I don't think this makes sense. Therefore here we first check that this is a server we care about before doing any validation.